lib/net.js

```diff
@@ -223,11 +223,18 @@ function Socket(options) {
   else if (options === undefined)
     options = {};
 
-  stream.Duplex.call(this, options);
+  // There is already a "no-half-open enforcer" in place so there is no need to
+  // inherit it again from `Duplex`.
+  stream.Readable.call(this, options);
+  stream.Writable.call(this, options);
```

Comment:
I am not really a fan of this … not calling the `Duplex` constructor seems like a bad idea in general, and it’s going to be an extra obstacle while working on merging all `StreamBase` JS wrappers…

Can we instead make the `Duplex` constructor not attach the event handler if `options.allowHalfOpen` is set? That would help all other Duplex streams as well…

Reply:
Yes, after https://github.com/nodejs/node/pull/18953 my idea was to override the `allowHalfOpen` option when calling the `Duplex` constructor in order to never inherit the listener but to do that the options object has to be copied. This was done for performance reasons.

Reply:
I'm open to suggestions.

Reply:
@lpinca Yeah, good point.

I don’t think we can get around that without making some kind of change to the `Duplex` API, but I would be okay with that.

The best I can come up with right now would be this:

- Add
  ```js
  Duplex.kHasHalfOpenEnforcer = Symbol('kHasHalfOpenEnforcer'); // or non-enumerable
  Duplex.prototype[Duplex.kHasHalfOpenEnforcer] = true;
  net.Socket.prototype[Duplex.kHasHalfOpenEnforcer] = false;
  ```
- Check for that property in the `Duplex` constructor

Another approach might be to create an internal `Duplex` constructor that omits the listener addition, which could serve as a base class of the “real”, publicly exposed `Duplex` stream…

Reply:
I'm not very happy with option 1, adapting the base class to the needs of the child seems kinda hackish. Will think a bit about the second approach.